File: Week7_DFS/validateBST.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */

class Solution {
    public boolean validateHelper(TreeNode node, Integer min, Integer max) {
        /* Tried using Integer.MAX_VALUE to no avail */
        if (node == null)
            return true;

        if ((min != null && node.val <= min) || (max != null && node.val >= max))
            return false;

        return (validateHelper(node.left, min, node.val) && validateHelper(node.right, node.val, max));
    }

    public boolean isValidBST(TreeNode root) {
        return validateHelper(root, null, null);
    }
}

File: Week7_DFS/validateBST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def validateHelper(self, node: TreeNode, lowBound, highBound) -> bool:
        if not node:
            return True
        
        if (lowBound != None and node.val <= lowBound or 
            highBound != None and node.val >= highBound):
            return False
        
        validateLeft = self.validateHelper(node.left, lowBound, node.val)
        validateRight = self.validateHelper(node.right, node.val, highBound)
        
        return validateLeft and validateRight
    
    def isValidBST(self, root: TreeNode) -> bool:
        return self.validateHelper(root, None, None)